fe/src/main/java/org/apache/doris/planner/MaterializedViewSelector.java

```diff
@@ -0,0 +1,453 @@
+    private Map<String, Set<String>> columnNamesInQueryOutput = Maps.newHashMap();
+
+    private boolean disableSPJGView;
+    private String reasonOfDisable;
+    private boolean isPreAggregation = true;
+
+    public MaterializedViewSelector(SelectStmt selectStmt, Analyzer analyzer) {
+        this.selectStmt = selectStmt;
+        this.analyzer = analyzer;
+        init();
+    }
+
+    /**
+     * There are two stages to choosing the best MV.
+     * Phase 1: Predicates
+     * According to aggregation and column information in the select stmt,
+     * the candidate MVs that meets the query conditions are selected.
+     * Phase 2: Priorities
+     * According to prefix index and row count in candidate MVs,
+     * the best MV is selected.
+     *
+     * @param scanNode
+     * @return
+     */
+    public void selectBestMV(ScanNode scanNode) throws UserException {
+        long start = System.currentTimeMillis();
+        Preconditions.checkState(scanNode instanceof OlapScanNode);
+        OlapScanNode olapScanNode = (OlapScanNode) scanNode;
+        Map<Long, List<Column>> candidateIndexIdToSchema = predicates(olapScanNode);
+        long bestIndexId = priorities(olapScanNode, candidateIndexIdToSchema);
+        LOG.info("The best materialized view is {} for scan node {} in query {}, cost {}",
+                 bestIndexId, scanNode.getId(), selectStmt.toSql(), (System.currentTimeMillis() - start));
+        olapScanNode.updateScanRangeInfo(bestIndexId, isPreAggregation, reasonOfDisable);
+    }
+
+    private Map<Long, List<Column>> predicates(OlapScanNode scanNode) {
+        // Step1: all of predicates is compensating predicates
+        Map<Long, List<Column>> candidateIndexIdToSchema = scanNode.getOlapTable().getVisibleIndexes();
+        OlapTable table = scanNode.getOlapTable();
+        Preconditions.checkState(table != null);
+        String tableName = table.getName();
+        // Step2: check all columns in compensating predicates are available in the view output
+        checkCompensatingPredicates(columnNamesInPredicates.get(tableName), candidateIndexIdToSchema);
+        // Step3: group by list in query is the subset of group by list in view or view contains no aggregation
+        checkGrouping(columnNamesInGrouping.get(tableName), candidateIndexIdToSchema);
+        // Step4: aggregation functions are available in the view output
+        checkAggregationFunction(aggregateColumnsInQuery.get(tableName), candidateIndexIdToSchema);
+        // Step5: columns required to compute output expr are available in the view output
+        checkOutputColumns(columnNamesInQueryOutput.get(tableName), candidateIndexIdToSchema);
+        // Step6: if table type is aggregate and the candidateIndexIdToSchema is empty,
+        if (table.getKeysType() == KeysType.AGG_KEYS && candidateIndexIdToSchema.size() == 0) {
+            // the base index will be added in the candidateIndexIdToSchema.
```

Comment:
Because we currently support query on aggregate tables directly.
Such as 
Aggregate table A (c1, c2, c3) . c1 and c2 is key while c3 is value. The aggregate function of c3 is sum.
The query is `select c1, c2, c3 from A`. The query equals `select c1, c2, sum(c3) from A group by c1, c2`.
However, selector filters out all indexes including the base index of table.
The reason is that base index is `select c1 ,c2 ,sum(c3) from A`. It could not be selected.
For selector, only origin data like A' (c1,c2,c3) while c3 is key also, could match those condition.
So here, we need special treatment for aggregate tables.